_inc/search/components/search-widget.jsx

```diff
@@ -0,0 +1,78 @@
+/** @jsx h */
+
+/**
+ * External dependencies
+ */
+import { h, Component } from 'preact';
+import Portal from 'preact-portal';
+import { debounce } from 'lodash';
+
+class SearchWidget extends Component {
+	constructor() {
+		super( ...arguments );
+		this.requestId = 0;
+		this.state = {
+			query: this.props.initialValue,
+			results: [],
+		};
+		this.onChangeQuery = this.onChangeQuery.bind( this );
+		this.getResults = debounce( this.getResults.bind( this ), 500 );
+		this.getResults( this.props.initialValue );
+	}
+	componentDidMount() {
+		if ( this.props.grabFocus ) {
+			this.input.focus();
+		}
+	}
+
+	bindInput = input => ( this.input = input );
+	onChangeQuery = event => {
+		const query = event.target.value;
+		this.setState( { query } );
+		this.getResults( query );
+	};
+
+	getResults( query ) {
+		if ( query ) {
+			const { api } = this.props;
+			if ( api ) {
+				this.requestId++;
+				const requestId = this.requestId;
+				api
+					.fetch( query )
+					.then( response => {
+						return response.json();
+					} )
+					.then( json => {
+						if ( this.requestId === requestId ) {
```

Comment:
Can you elaborate on why this mechanism? Wouldn't `debounce` protect only latest request running? Or if not, should you implement cancellation so that you won't have multiple requests on fly simultaneously?

Reply:
I've seen folks think that `debounce` protects against this, and it doesn't. Cancellation of previous fetches would probably be easier to follow though.

Slow connections and unreliable connections are the reason you can't rely on `debounce` to protect against race conditions. If request 1 fires off, and then half a second later (a typical debounce timeout) request 2 fires off, there's no guarantee they'll resolve in that order. On fast networks they _probably_ would, but on mobile and other networks that are quite variable in speed, there's no guarantee.